### PR TITLE
add read-only setting for resource plan configurable features

### PR DESCRIPTION
**Pull request by davidleger95**
## Reason for change

Prevent inputs from being changed when viewing a resource plan. Only allow editing when resizing a plan. 

## Testing

These changes will require thorough testing of the plan selector, creating/resizing a resource, and vieing resource details. 

## Checklist

- [ ] **CHANGELOG**: The **Unreleased** section of CHANGELOG was updated
- [ ] **Prop changes**: [docs][docs] were updated
- [ ] **Prop changes**: E2E tests were updated, testing from the highest level possible
- [ ] **Platform testing**: If this change should be [tested against a platform][platform-testing], has it been?

[docs]: https://ui.sandbox.manifold.co
[platform-testing]: https://app.gitbook.com/@manifold/s/engineering/playbooks/testing-your-code-against-a-platform

### CODE REVIEW

File: src/components/manifold-select/manifold-select.tsx [hunk 1]
```diff
@@ -32,7 +33,12 @@ export class ManifoldSelect {
   @logger()
   render() {
     return (
-      <select name={this.name} required={this.required} onChange={this.onChangeHandler}>
+      <select
+        name={this.name}
+        required={this.required}
+        onChange={this.onChangeHandler}
+        disabled={this.disabled}
```
Comment (davidleger95) [inline src/components/manifold-select/manifold-select.tsx]: One question I have about this is do we want to disable the inputs (as I'm doing here) or display the configured features as if they were fixed?

Comment (sslotsky) [reply src/components/manifold-select/manifold-select.tsx]: Is the problem here that this component is used in two different contexts? i.e. we use it in the plan selector but we also use it by itself to display the current features?

The question you're asking here is a design question, but my instinct says that it should display as if it were fixed. And if that's the case, then I think that `manifold-resource-plan` becomes a little bit kludgy, because there will be multiple instances where it has to display one component if it's read only and another component otherwise. In that case I'd almost prefer it if `manifold-plan-details` displayed everything fixed, and then the plan selector displayed a different component. What do you think?

Comment (dangodev) [reply src/components/manifold-select/manifold-select.tsx]: I could go either way but I think I like the disabled inputs better. This has an affordance that “you _can_ edit these in certain situations; just not now.” It communicates more is possible to the user w/o saying anything.

File: src/components/manifold-active-plan/manifold-active-plan.tsx [hunk 2]
```diff
@@ -64,6 +66,7 @@ export class ManifoldActivePlan {
         resourceRegion={resourceRegion}
         scrollLocked={true}
         region={this.selectedResource && this.selectedResource.region}
+        configuredFeatures={formatGatewayFeatures(this.selectedResource?.configuredFeatures?.edges)}
```
Comment (dangodev) [inline src/components/manifold-active-plan/manifold-active-plan.tsx]: This is something to watch—TypeScript doesn’t transform these, so this will break most non-bleeding-edge browsers. Check the output JS to see if these still exist or not.

File: src/components/manifold-plan-selector/resource.graphql [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+
+          ... on BooleanConfiguredFeature {
+            booleanValue: value
+          }
+          ... on NumberConfiguredFeature {
+            numberValue: value
+          }
+          ... on StringConfiguredFeature {
+            stringValue: value
+          }
```
Comment (davidleger95) [inline src/components/manifold-plan-selector/resource.graphql]: Problem here: we can't generate types from graphql with this :/

File: src/components/manifold-plan-details/manifold-plan-details.tsx [hunk 4]
```diff
@@ -61,15 +65,21 @@ export class ManifoldPlanDetails {
...
+  @Watch('configuredFeatures') configuredFeaturesChange(configuredFeatures: Gateway.FeatureMap) {
+    if (this.plan) {
+      this.resetFeatures(this.plan, configuredFeatures);
+    }
+  }
+
   @loadMark()
   @loadMark()
-  componentWillLoad() {
+  componentDidLoad() {
```
Comment (dangodev) [inline src/components/manifold-plan-details/manifold-plan-details.tsx]: Why the change here?

File: src/components/manifold-resource-plan/manifold-resource-plan.tsx [hunk 5]
```diff
@@ -26,14 +27,15 @@ export class ManifoldResourcePlan {
       );
     }
 
-    // TODO: remove `as` and update types
```
Comment (dangodev) [inline src/components/manifold-resource-plan/manifold-resource-plan.tsx]: This still needs to stay until the `as` is removed (we should be using generated types, but we need to finish converting `manifold-plan-details` first)

Comment (dangodev) [reply src/components/manifold-resource-plan/manifold-resource-plan.tsx]: Or if the `as` can go it probably should!

File: src/components/manifold-plan-selector/resource.graphql (hunk 3 above)
Comment (dangodev) [reply src/components/manifold-plan-selector/resource.graphql]: Does this work? https://graphql-code-generator.com/docs/plugins/fragment-matcher#__docusaurus

Comment (dangodev) [reply src/components/manifold-plan-selector/resource.graphql]: Actually nevermind—this is for something different.

I don’t think codegen is the problem; I think our GraphQL server is the problem. This should be able to work! But our server needs to have `fragment`s in its schema, and it doesn’t.

Our GraphQL schema should have something like:

```graphql
fragment StringConfiguredFeature on Resource {
  …
}
fragment NumberConfiguredFeature on Resource {
  …
}
…
```

See [this issue](https://github.com/dotansimha/graphql-code-generator/pull/2540) as an example

Comment (davidleger95) [reply src/components/manifold-plan-selector/resource.graphql]: fragments are on the client though, aren't they? We use fragments for the plan query already: https://github.com/manifoldco/ui/blob/1f053ae1120f21c735afd9dbee20f6b8a8ff23e5/src/components/manifold-plan/plan.graphql

If that works then this should also work, but it appears the types aren't being exposed by the server. however, they are exposed in GraphQL Playground

File: src/components/manifold-resource-plan/manifold-resource-plan.tsx (hunk 5 above)
Comment (davidleger95) [reply src/components/manifold-resource-plan/manifold-resource-plan.tsx]: ah, oops i'll put it back

File: src/components/manifold-plan-selector/resource.graphql (hunk 3 above)
Comment (davidleger95) [reply src/components/manifold-plan-selector/resource.graphql]: Update: Sam made me realize the changes to the API weren't deployed to prod 🤦‍♂ it all works fine haha

Comment (dangodev) [reply src/components/manifold-plan-selector/resource.graphql]: That’s good news!

File: src/components/manifold-plan-details/manifold-plan-details.tsx (hunk 4 above)
Comment (davidleger95) [reply src/components/manifold-plan-details/manifold-plan-details.tsx]: was just trying testing things out

File: src/components/manifold-plan-details/manifold-plan-details.tsx [hunk 6]
```diff
@@ -61,13 +65,12 @@ export class ManifoldPlanDetails {
...
     // update features
-    this.resetFeatures(newPlan);
+    this.resetFeatures(newPlan, this.configuredFeatures);
   }
   @Watch('resourceRegion') resourceRegionChange(newRegion: string) {
     this.regionId = newRegion;
   }
 
-  @loadMark()
   @loadMark()
```
Comment (davidleger95) [inline src/components/manifold-plan-details/manifold-plan-details.tsx]: Any reason why there were 2 of these here?

Comment (sslotsky) [reply src/components/manifold-plan-details/manifold-plan-details.tsx]: Not a good one, no doubt :grinning:

File: src/utils/configuredFeatures.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
+import { Gateway } from 'types/gateway';
+
+export interface ConfiguredFeatureGraphQL {
```
Comment (sslotsky) [inline src/utils/configuredFeatures.ts]: Is there a way to get this type from the resource query definition in graphql.ts? Something like `ResourceQuery['configuredFeatures']`?

File: src/utils/configuredFeatures.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+export interface ConfiguredFeature {
+  label: string;
+  value?: string;
+}
+
+const isDefined = (x: unknown) => typeof x !== 'undefined';
+
+export const formatConfiguredFeatures = (
```
Comment (sslotsky) [inline src/utils/configuredFeatures.ts]: Doesn't look like this one is used?

File: src/utils/configuredFeatures.ts (hunk 7 above)
Comment (davidleger95) [reply src/utils/configuredFeatures.ts]: yes, i think so. had this here originally because i couldn't generate types for it. but now i can :)

File: src/utils/configuredFeatures.ts (hunk 8 above)
Comment (davidleger95) [reply src/utils/configuredFeatures.ts]: I'll be used in my next PR, I can take it out for now if you'd like though. what do you think?

Comment (sslotsky) [reply src/utils/configuredFeatures.ts]: Yeah I'd say remove it for now, bring it back when you're ready to use it.

File: src/components/manifold-active-plan/manifold-active-plan.tsx (hunk 2 above)
Comment (davidleger95) [reply src/components/manifold-active-plan/manifold-active-plan.tsx]: I checked the outputs (esm, esm-es5, cjs, collection) and they do transpile to ternaries

File: src/utils/configuredFeatures.ts (hunk 8 above)
Comment (davidleger95) [reply src/utils/configuredFeatures.ts]: removed

File: package.json [hunk 9]
```diff
@@ -77,7 +77,7 @@
     },
     {
       "path": "./dist/esm/manifold-number-input*.js",
-      "maxSize": "36 kB",
+      "maxSize": "40 kB",
```
Comment (davidleger95) [inline package.json]: Had to increase this limit. Seems like this module is getting quite large in comparison to others though. 🤔

Comment (dangodev) [reply package.json]: That’s fine. This special little chunk is several components in one to save requests. That’s why it’s a separate size benchmark from the others—so it can be bumped on its own without affecting any others.
